robot-detect

```diff
@@ -34,20 +34,40 @@ MSG_FASTOPEN = 0x20000000
 EXECUTE_BLINDING = True
 
 
-def get_rsa_from_server(server, port):
+def create_socket(ip_family=4, timeout=None):
+    if ip_family == 4:
+        s = socket.socket()
+        if timeout is not None:
+            s.settimeout(timeout)
+        return s
+    s = socket.socket(socket.AF_INET6)
+    if timeout is not None:
+        s.settimeout(timeout)
+    return s
+
+
+def get_rsa_from_server(server, ip_family):
     try:
         ctx = ssl.create_default_context()
         ctx.check_hostname = False
         ctx.verify_mode = ssl.CERT_NONE
         ctx.set_ciphers("RSA")
-        raw_socket = socket.socket()
-        raw_socket.settimeout(timeout)
+        raw_socket = create_socket(ip_family, timeout)
         s = ctx.wrap_socket(raw_socket)
-        s.connect((server, port))
-        cert_raw = s.getpeercert(binary_form=True)
-        cert_dec = x509.load_der_x509_certificate(cert_raw, default_backend())
-        s.close()
-        return cert_dec.public_key().public_numbers().n, cert_dec.public_key().public_numbers().e
+        try:
+            s.connect(server)
+            cert_raw = s.getpeercert(binary_form=True)
+            cert_dec = x509.load_der_x509_certificate(cert_raw, default_backend())
+            s.close()
+            return (
+                cert_dec.public_key().public_numbers().n,
+                cert_dec.public_key().public_numbers().e,
+            )
+        except OSError as e:
+            print("Error: %s \nYou are probably trying to connect to an IPv6 service via IPv4 or vice-versa. "
```

Comment:
Actually, this error could be due to a different issue, right?  It might be better to keep it generic.  The user should figure out why they couldn't connect.

Also, an IP family mismatch should be very rare if the first address from `socket.getaddrinfo` is used, because it should return the IPv4 first for IPv4-only clients.

Reply:
Yes it could. From my tests it seems like some other exception is thrown before if there are other problems, so in this case I don't see many other possibilities for the error to appear if not for the reported one. Still, I will be more generic.

Yes, the mismatch is very rare for that reason.